File: tests/golden/sandbox/recipes/gt1_happy.py
"""GT-1 — happy path: add ``multiply`` to the toy calculator.

Five cycles total:

1. plan halt              → recipe sets ``state.plan``
2. apply-plan halt        → recipe writes source + test, sets ``state.changes``
3. run-tests halt         → recipe runs pytest, sets ``state.tests`` (success)
4. review-changes halt    → recipe sets ``state.verify`` (success)
5. report runs            → engine exits 0 with the delivery report
"""
from __future__ import annotations

from pathlib import Path
from typing import Any

from . import _helpers as h

META = {
    "gt_id": "GT-1",
    "ticket_relpath": "tickets/gt-1-happy.json",
    "persona": None,
    "cycle_cap": 6,
}

_MULTIPLY_SRC = '''

def multiply(a: int, b: int) -> int:
    return a * b
'''

_MULTIPLY_TEST = '''

def test_multiply_returns_product() -> None:
    from src.calculator import multiply
    assert multiply(3, 4) == 12
    assert multiply(0, 5) == 0
    assert multiply(-2, 3) == -6
'''


def build_recipe(workspace: Path) -> dict[str, Any]:
    """Return the directive→step mapping with ``workspace`` bound in."""

    def on_create_plan(state: dict[str, Any], record) -> dict[str, Any]:
        state["plan"] = h.standard_plan(
            "Add multiply(a, b)",
            "Append multiply(a: int, b: int) -> int to src/calculator.py",
            "Append test_multiply_returns_product to tests/test_calculator.py",
            "Run pytest to confirm the new test passes",
        )
        record.recipe_notes.append("plan recorded with 3 steps")
        return state

    def on_apply_plan(state: dict[str, Any], record) -> dict[str, Any]:
        h.append_to_file(workspace, "src/calculator.py", _MULTIPLY_SRC)
        h.append_to_file(workspace, "tests/test_calculator.py", _MULTIPLY_TEST)
        state["changes"] = h.base_changes(
            "src/calculator.py",
            "tests/test_calculator.py",
        )
        record.recipe_notes.append("appended multiply + test_multiply_returns_product")
        return state

    def on_run_tests(state: dict[str, Any], record) -> dict[str, Any]:
        state["tests"] = h.run_pytest(workspace)
        record.recipe_notes.append(
            f"pytest verdict={state['tests']['verdict']}",
        )
        return state

    def on_review_changes(state: dict[str, Any], record) -> dict[str, Any]:
        state["verify"] = h.simulated_review_verdict()
        record.recipe_notes.append("review-changes simulated success")
        return state

    return {
        "create-plan": on_create_plan,
        "apply-plan": on_apply_plan,
        "run-tests": on_run_tests,
        "review-changes": on_review_changes,
    }


__all__ = ["META", "build_recipe"]

File: tests/golden/sandbox/recipes/gt1_happy.ts
/**
 * GT-1 — happy path: add `multiply` to the toy calculator (TS twin of the
 * retired gt1_happy.py; the injected source/test are now TypeScript).
 *
 * Five cycles total:
 *   1. create-plan halt   → recipe sets state.plan
 *   2. apply-plan halt    → recipe writes source + test, sets state.changes
 *   3. run-tests halt     → recipe runs vitest, sets state.tests (success)
 *   4. review-changes halt→ recipe sets state.verify (success)
 *   5. report runs        → engine exits 0 with the delivery report
 */
import type { RecipeModule } from './index.js';
import * as h from './_helpers.js';
import type { CycleRecord, Dict, RecipeStep } from '../runner.js';

export const META: RecipeModule['META'] = {
    gt_id: 'GT-1',
    ticket_relpath: 'tickets/gt-1-happy.json',
    persona: null,
    cycle_cap: 6,
};

const MULTIPLY_SRC = `

export function multiply(a: number, b: number): number {
    return a * b;
}
`;

const MULTIPLY_TEST = `

import { multiply } from '../src/calculator.js';

it('multiply returns product', () => {
    expect(multiply(3, 4)).toBe(12);
    expect(multiply(0, 5)).toBe(0);
    expect(multiply(-2, 3)).toBe(-6);
});
`;

export function buildRecipe(workspace: string): Record<string, RecipeStep> {
    const onCreatePlan: RecipeStep = (state: Dict, record: CycleRecord) => {
        state['plan'] = h.standard_plan(
            'Add multiply(a, b)',
            'Append multiply(a: number, b: number): number to src/calculator.ts',
            'Append a multiply test to tests/calculator.test.ts',
            'Run vitest to confirm the new test passes',
        );
        record.recipe_notes.push('plan recorded with 3 steps');
        return state;
    };

    const onApplyPlan: RecipeStep = (state: Dict, record: CycleRecord) => {
        h.append_to_file(workspace, 'src/calculator.ts', MULTIPLY_SRC);
        h.append_to_file(workspace, 'tests/calculator.test.ts', MULTIPLY_TEST);
        state['changes'] = h.base_changes('src/calculator.ts', 'tests/calculator.test.ts');
        record.recipe_notes.push('appended multiply + multiply test');
        return state;
    };

    const onRunTests: RecipeStep = (state: Dict, record: CycleRecord) => {
        state['tests'] = h.run_vitest(workspace);
        record.recipe_notes.push(`vitest verdict=${(state['tests'] as Dict)['verdict'] as string}`);
        return state;
    };

    const onReviewChanges: RecipeStep = (state: Dict, record: CycleRecord) => {
        state['verify'] = h.simulated_review_verdict();
        record.recipe_notes.push('review-changes simulated success');
        return state;
    };

    return {
        'create-plan': onCreatePlan,
        'apply-plan': onApplyPlan,
        'run-tests': onRunTests,
        'review-changes': onReviewChanges,
    };
}

const recipe: RecipeModule = { META, buildRecipe };
export default recipe;
